src/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs

```diff
@@ -13,34 +15,64 @@ public sealed partial class SafePipeHandle : SafeHandle
     {
         private const int DefaultInvalidHandle = -1;
 
-        /// <summary>Opens the specified file with the requested flags and mode.</summary>
-        /// <param name="path">The path to the file.</param>
-        /// <param name="flags">The flags with which to open the file.</param>
-        /// <param name="mode">The mode for opening the file.</param>
-        /// <returns>A SafeFileHandle for the opened file.</returns>
-        internal static SafePipeHandle Open(string path, Interop.Sys.OpenFlags flags, int mode)
+        // For anonymous pipes, SafePipeHandle.handle is the file descriptor of the pipe, and the 
+        // _named* fields remain null. For named pipes, SafePipeHandle.handle is a copy of the file descriptor 
+        // extracted from the Socket's SafeHandle, and the _named* fields are the socket and its safe handle.
+        // This allows operations related to file descriptors to be performed directly on the SafePipeHandle,
+        // and operations that should go through the Socket to be done via _namedPipeSocket.  We keep the
+        // Socket's SafeHandle alive as long as this SafeHandle is alive.
+
+        private Socket _namedPipeSocket;
+        private SafeHandle _namedPipeSocketHandle;
+
+        internal SafePipeHandle(Socket namedPipeSocket) : base((IntPtr)DefaultInvalidHandle, ownsHandle: true)
         {
-            // Ideally this would be a constrained execution region, but we don't have access to PrepareConstrainedRegions.
-            SafePipeHandle handle = Interop.CheckIo(Interop.Sys.OpenPipe(path, flags, mode));
+            Debug.Assert(namedPipeSocket != null);
+            _namedPipeSocket = namedPipeSocket;
+
+            // TODO: Issue https://github.com/dotnet/corefx/issues/6807
+            // This is unfortunately the only way of getting at the Socket's file descriptor right now, until #6807 is implemented.
+            _namedPipeSocketHandle = (SafeHandle)typeof(Socket).GetTypeInfo().GetDeclaredProperty("SafeHandle")?.GetValue(namedPipeSocket, null);
```

Comment:
I can do that, or at least cache it into a static on first use.  I was sort-of hoping this would go away before it really mattered, but better safe-than-sorry. :smile: